src/main/java/org/kaazing/netx/ws/internal/io/WsMessageReader.java

```diff
@@ -404,6 +455,110 @@ private synchronized int readText(char[] cbuf, int offset, int length) throws IO
         return offset - mark;
     }
 
+    private void filterControlFrames() throws IOException {
+        int opcode = header[0] & 0x0F;
+
+        if ((opcode == 0x00) || (opcode == 0x01) || (opcode == 0x02)) {
+            return;
+        }
+
+        readPayloadLength();
+
+        switch (opcode) {
+        case 0x08:
+            int code = 0;
+            byte[] reason = null;
+
+            if (payloadLength >= 2) {
+                // Read the first two bytes as the CLOSE code.
+                int b1 = in.read();
+                int b2 = in.read();
+
+                code = ((b1 & 0xFF) << 8) | (b2 & 0xFF);
+                if ((code == 1005) || (code == 1006) || (code == 1015)) {
+                    code = 1002;
+                }
+
+                // If reason is also received, then just drain those bytes.
+                if (payloadLength > 2) {
+                    reason = new byte[(int) (payloadLength - 2)];
+                    int bytesRead = in.read(reason);
+
+                    if (bytesRead == -1) {
+                        throw new IOException("End of stream");
+                    }
+
+                    if ((reason.length > 123) || !Utf8Util.isValidUTF8(reason)) {
+                        code = 1002;
+                    }
+
+                    if (code != 1000) {
+                        reason = null;
+                    }
+                }
+            }
+
+            if (out.wasCloseSent()) {
```

Comment:
Suggest modeling `wasCloseSent()` as connection state enum on `WsURLConnectionImpl` since we already have a reference, rather than coupling the input stream and output stream together directly.  What do you think?